File: extensions/jetbrains/src/main/kotlin/com/deecency/udara/cli/PubspecVersion.kt
package com.deecency.udara.cli

import java.io.File

/**
 * Reads and rewrites the top-level `version:` line of pubspec.yaml. Mirrors
 * extensions/vscode/src/pubspecVersion.ts; keep the two in sync.
 */
object PubspecVersion {
    /** Flutter's format: build-name with an optional +build-number. */
    val PATTERN = Regex("""^\d+\.\d+\.\d+(?:[-.][0-9A-Za-z.-]+)?(?:\+\d+)?$""")

    private val LINE = Regex("""^(version:[ \t]*)(["']?)([^"'\s#]+)\2(.*)$""", RegexOption.MULTILINE)

    fun file(root: File) = File(root, "pubspec.yaml")

    fun read(root: File): String? {
        val f = file(root)
        if (!f.isFile) return null
        return LINE.find(f.readText())?.groupValues?.get(3)
    }

    /**
     * "1.4.0" on a pubspec at "1.3.9+41" becomes "1.4.0+41" so the store
     * build number is kept unless one is given. Blank input means no override.
     */
    fun resolve(input: String, current: String?): String? {
        val trimmed = input.trim()
        if (trimmed.isEmpty()) return null
        if (trimmed.contains('+') || current == null || !current.contains('+')) return trimmed
        return "$trimmed+${current.substringAfter('+')}"
    }

    fun replaceIn(content: String, version: String): String {
        val m = LINE.find(content) ?: error("pubspec.yaml has no \"version:\" line to override.")
        val (prefix, quote, _, rest) = m.destructured
        return content.replaceRange(m.range, "$prefix$quote$version$quote$rest")
    }

    fun write(root: File, version: String) {
        val f = file(root)
        f.writeText(replaceIn(f.readText(), version))
    }
}

File: extensions/vscode/src/pubspecVersion.ts
import * as fs from 'fs';
import * as path from 'path';

/** Flutter's pubspec version format: build-name with an optional +build-number. */
export const VERSION_PATTERN = /^\d+\.\d+\.\d+(?:[-.][0-9A-Za-z.-]+)?(?:\+\d+)?$/;

const VERSION_LINE = /^(version:[ \t]*)(["']?)([^"'\s#]+)\2(.*)$/m;

export function pubspecPath(root: string): string {
  return path.join(root, 'pubspec.yaml');
}

/** Reads the `version:` value from pubspec.yaml, or undefined when absent. */
export function readPubspecVersion(root: string): string | undefined {
  const file = pubspecPath(root);
  if (!fs.existsSync(file)) {
    return undefined;
  }
  return VERSION_LINE.exec(fs.readFileSync(file, 'utf8'))?.[3];
}

/**
 * Resolves what the user typed into the version to write. A version without
 * a build number keeps the current one, so "1.4.0" on "1.3.9+41" becomes
 * "1.4.0+41". Returns undefined for an empty input (keep the pubspec value).
 */
export function resolveVersion(input: string, current: string | undefined): string | undefined {
  const trimmed = input.trim();
  if (!trimmed) {
    return undefined;
  }
  if (trimmed.includes('+') || !current?.includes('+')) {
    return trimmed;
  }
  return `${trimmed}+${current.split('+')[1]}`;
}

/** Rewrites only the value of the `version:` line, keeping quotes and comments. */
export function replaceVersionInContent(content: string, version: string): string {
  if (!VERSION_LINE.test(content)) {
    throw new Error('pubspec.yaml has no "version:" line to override.');
  }
  return content.replace(
    VERSION_LINE,
    (_m, prefix: string, quote: string, _old: string, rest: string) =>
      `${prefix}${quote}${version}${quote}${rest}`,
  );
}

export function writePubspecVersion(root: string, version: string): void {
  const file = pubspecPath(root);
  fs.writeFileSync(file, replaceVersionInContent(fs.readFileSync(file, 'utf8'), version));
}
